tests/linearizability/model.go

```diff
@@ -187,24 +198,32 @@ func stepDelete(state EtcdState, request EtcdRequest, response EtcdResponse) (bo
 		state.FailedWrite = &request
 		return true, state
 	}
-	// revision should never decrease
-	if response.Revision < state.LastRevision {
-		return false, state
-	}
 	deleteSucceeded := response.Deleted != 0
 	keySet := state.Value != ""
 
-	// non-existent key cannot be deleted.
-	if deleteSucceeded != keySet && state.FailedWrite == nil {
-		return false, state
-	}
-	//if key was deleted, response revision should increase
-	if deleteSucceeded && (response.Revision != state.LastRevision+1 || !keySet) && (state.FailedWrite == nil || response.Revision < state.LastRevision+2) {
-		return false, state
-	}
-	//if key was not deleted, response revision should not change
-	if !deleteSucceeded && state.LastRevision != response.Revision && state.FailedWrite == nil {
-		return false, state
+	if state.FailedWrite == nil {
+		// non-existent key cannot be deleted.
+		if deleteSucceeded != keySet {
+			return false, state
+		}
+		// if key was deleted, response revision should increase by one
+		if deleteSucceeded && response.Revision != state.LastRevision+1 {
+			return false, state
+		}
+		// if key was not deleted, response revision should not change
+		if !deleteSucceeded && state.LastRevision != response.Revision {
+			return false, state
+		}
+	} else {
+		// if key was set, revision must increase as either key was deleted or failed request was persisted.
+		if keySet && response.Revision < state.LastRevision+1 {
+			return false, state
+		}
```

Comment:
This may not be correct. Let's work with an example, assuming there are three successive requests:
1. read: successful, so the state.Value is set as the `response.GetData`
2. delete:  failed on client, but the etcdserver actually successfully deleted the k/v. So the state.Value keeps unchanged, but `state.FailedWrite` is updated 
3. delete: Since the k/v has already been deleted, so the revision will not increase.

Reply:
It will increase due to delete in point 2. 

Assuming that we know that at least one write failed `state.FailedWrite != nil`, table of cases is:

keySet | deleteSucceeded | revision change | comment
-----------|--------------------------|-------------------------|-----------------
 true    |  true                        | +1                         | Delete just worked on existing key. Failed request were lost
 false    |  true                        | +2                         | Delete succeeded on non-existing key. Means that there was at least one Put request that failed. Meaning since last state update there were two transactions: Put, Delete(success).
 true    |  false                        | +1                         | Delete failed on existing key. Means that there was at least one Put request that failed. Meaning since last state update there were two transactions: Put, Delete(failed).
 false    |  false                        | 0                         | Delete just failed on non-existing key. Failed request were lost.

Reply:
Anyway, please see https://github.com/etcd-io/etcd/pull/14902#issuecomment-1340041130